compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/optimizer/LargeMethodOptimizer.java

```diff
@@ -104,28 +122,1063 @@ public void splitLargeBIRFunctions(BIRPackage birPkg) {
+        // moving instructions and terminators until this one (including this) to a new function.
+        // Criteria for choosing split point is as follows. We go from bottom to top searching for array value operands
+        // in both rhs and lhs. If we find one, that is a split point, unless we find that previously in the same BB.
+        // A new flag is added to the split point to depict whether to split the function here (splitHere).
+        // If it is false only array populating instructions are added.
+        Map<BIRAbstractInstruction, SplitPointDetails> insSplitPoints = new HashMap<>();
+        for (int bbIndex = bbs.size() - 2; bbIndex >= 0; bbIndex--) {
+            Set<BIROperand> operandsInSameBB = new HashSet<>();
+            BIRBasicBlock bb = bbs.get(bbIndex);
+            List<BIRNonTerminator> bbInstructions = bb.instructions;
+            int lastInsNum = getLastInsNumAndHandleTerminator(bbs, arrayValuesOperands, remainingArrayValueOperands,
+                    insSplitPoints, bbIndex, operandsInSameBB, bb, bbInstructions);
+            for (int insIndex = lastInsNum; insIndex >= 0; insIndex--) {
+                populateInsSplitPoints(arrayValuesOperands, insSplitPoints, bbInstructions.get(insIndex),
+                        operandsInSameBB, remainingArrayValueOperands);
+            }
+        }
+        return insSplitPoints;
+    }
+
+    private void populateInsSplitPoints(Set<BIROperand> arrayValuesOperands,
+                                        Map<BIRAbstractInstruction, SplitPointDetails> insSplitPoints,
+                                        BIRAbstractInstruction ins, Set<BIROperand> operandsInSameBB,
+                                        Set<BIROperand> remainingArrayValueOperands) {
+        BIROperand insLhsOp = ins.lhsOp;
+        if (insLhsOp != null) {
+            addOperandToInsSplitPoints(arrayValuesOperands, insSplitPoints, ins, insLhsOp, operandsInSameBB,
+                    remainingArrayValueOperands);
+        }
+    }
+
+    private void addOperandToInsSplitPoints(Set<BIROperand> arrayValuesOperands,
+                                            Map<BIRAbstractInstruction, SplitPointDetails> insSplitPoints,
+                                            BIRAbstractInstruction ins, BIROperand insOperand,
+                                            Set<BIROperand> operandsInSameBB,
+                                            Set<BIROperand> remainingArrayValueOperands) {
+        if (arrayValuesOperands.contains(insOperand) && !operandsInSameBB.contains(insOperand)) {
+            operandsInSameBB.add(insOperand);
+            if (insSplitPoints.containsKey(ins)) {
+                SplitPointDetails splitPointDetails = insSplitPoints.get(ins);
+                List<BIROperand> operandList = splitPointDetails.arrayElementsBIROperands;
+                operandList.add(insOperand);
```

Comment:
Some lines in the `addOperandToInsSplitPoints` are not covered. Is it expected?

Reply:
There can be a possibility for the same instruction (e.g. function call) to have multiple key, value operands of a map value. Hence both the LHS op and RHS ops of the same instruction can call this method multiple times and may have a chance to `(insSplitPoints.containsKey(ins))` become `true`. Since I am not certain this will never happen, I have kept the code for now. We can remove these lines later if we found that this can't happen. Currently the tests show in favour of removing, but can check further before removing.